### PR TITLE
Refactor MerkleTreeGadget implementation

### PR DESCRIPTION
**Pull request by mike1729**
## Description

There are two implementations of `MerkleTreeGadget` for `RescueMerkleTree` and `RescueSparseMerkleTree` that are equivalent except types they use. Recently a bug was found in the impl for `RescueMerkleTree` that affected also the other and only one was fixed.
This PR proposes a refactor to a common implementation for both types.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

**Comment by mike1729** [comment]
There is still some work to be done in moving some tests around but I wanted first to check if you like the direction of the change.

**Comment by mrain** [comment]
> There is still some work to be done in moving some tests around but I wanted first to check if you like the direction of the change.

Thanks very much for spotting this. It will be very helpful if you proceed ahead.
Much appreciated!

**Comment by mike1729** [comment]
Ready for review :)

File: primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs [hunk 1]
```diff
@@ -6,330 +6,3 @@
```
Comment (mrain) [inline primitives/src/circuit/merkle_tree/rescue_merkle_tree.rs:6]: Shoule delete this file?

**Comment by mrain** [comment]
LGTM in general, minor comments above.

Also do you think it's better to rename `sparse_merkle_tree.rs` as `universal_merkle_tree.rs`?

**Comment by mike1729** [comment]
@mrain addressed all comments :)

**Review by mrain — APPROVED** [review]
LGTM. Thanks a lot!

**Pull request merged**